### ISSUE TITLE
Applying on typical PyTorch data format

### ISSUE THREAD

**Issue by fshamsafar** [patrick-kidger/torchcubicspline#5]
Hi,

Is it possible to apply the cubic spline interpolation on one dimension of a data with the data format as in PyTorch `[BatchSize, Channel, Height, Width]`?

Thanks

**Comment by patrick-kidger** [comment]
That, is you want to interpolate along e.g. the height dimension? Sure:
```
data = torch.rand(batch, channel, height, width)
data_perm = data.transpose(1, 3)  # shape (batch, width, height, channel)
```
Then do interpolation. Batch and width both get treated as batch dimensions, and interpolation is performed along the height dimension. Once you've sampled from the cubic spline, transpose the dimensions back again.

**Comment by fshamsafar** [comment]
Actually, I want to do the interpolation along the channel dimension. Namely, the input data is `[B, C, H, W]` and I wish to get non-integer values at 1/4 integer values of channel dimension, `[B, C/4, H, W]`.
I am confused how I should apply the `natural_cubic_spline_coeffs` function after I permute the dimensions to `[B, H, C/4, W]`.

**Comment by patrick-kidger** [comment]
Along the channel dimension? Sounds a little odd, but sure.

`natural_cubic_spline_coeffs` takes in tensors of shape `(..., X, Y)`, corresponding to a (batch of) sequences of length `X`, with each sequence element being a vector of size `Y`. There's actually no difference between the `...` dimensions and the `Y` dimension; they're both batch dimensions as far as interpolation is concerned. This `Y` dimension is put at the end just as a convenience, to fit the common "batch-length-channel" data format.

So in this case, your channel dimension corresponds to `X`, and all of the other dimensions should be spread amongst the `...` or the `Y`. Which ones you want to put where are up to you. If you want to treat them all consistently you could permute them to `[B, H, W, C, 1]` with a dummy dimension at the end; or you could do `[B, H, C, W]` as you suggest. Either way just call directly `natural_cubic_spline_coeffs` on that.

Does that make sense?

**Comment by fshamsafar** [comment]
Great! Thanks!
So, basically but a snippet of code similar to the one below, I guess I will get all the data information at 1/4 of the channel dimension. Am I right?!

```
batch = 2
channel = 128
height = 512
width = 512
x = torch.rand(batch, height, channel, width)
t = torch.linspace(0, channel/4, channel)
coeffs = natural_cubic_spline_coeffs(t, x)
spline = NaturalCubicSpline(coeffs)
out = spline.derivative(t)
```

**Comment by patrick-kidger** [comment]
More or less. The final call to spline.derivative will be evaluated at all the t points, and those were also specified as being the times of the knots in natural_cubic_spline_coeffs, so the values you get out will be the derivative at all of your inputs x.

I don't know if that aligns with what you're after exactly.
Maybe create natural_cubic_spline_coeffs with torch.linspace(0, channels-1, channels), and then evaluating at torch.linspace(0, channels-1, 4*channels) is what you're after? That will evaluate at every quarter-channel.

**Comment by f-sh-f-sh** [comment]
Yes, I think with your hint, this is exactly what I want. As a final step, since I wish to have an output with the same size as the input, I need to take every four element. 

The output of the code below seems to be what I was looking for, I guess.

```
batch = 2
channel = 128
height = 512
width = 512

x = torch.rand(batch, height, channel, width)
print(x.size())
print(x[0,0,0:20,0],'\n')

t = torch.linspace(0, channel-1, channel)
coeffs = natural_cubic_spline_coeffs(t, x)
spline = NaturalCubicSpline(coeffs)
out = spline.evaluate(torch.linspace(0, channel-1, 4*channel))
print(out.size())
print(out[0,0,0:20,0],'\n')

out = out[:,:,::4,:]
print(out.size())
print(out[0,0,0:20,0],'\n')
```

Thanks.

**Comment by f-sh-f-sh** [comment]
Actually, what I needed was something similar to the interpolation in this [paper](https://openaccess.thecvf.com/content_WACV_2020/papers/Kallwies_Triple-SGM_Stereo_Processing_using_Semi-Global_Matching_with_Cost_Fusion_WACV_2020_paper.pdf), section 4.3.
